gtda/images/filtration.py

```diff
@@ -0,0 +1,160 @@
+
+        for i in range(Xh.shape[0]):
+            Xh[i][X[i] == True] = np.dot(self.mesh_[X[i] == True],
+                                         self.direction_).reshape((-1,))
+
+        return Xh
+
+    def fit(self, X, y=None):
+        """Calculate `n_dimensions_`, 'mesh_' and `max_value_` from a
+        collection of binary image. Then, return the estimator.
+
+        This method is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Input data. Each entry along axis 0 is interpreted as a 2D or 3D
+            grayscale image.
+
+        y : None
+            There is no need of a target in a transformer, yet the pipeline API
+            requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        X = check_array(X,  ensure_2d=False, allow_nd=True)
+
+        self.n_dimensions_ = len(X.shape) - 1
+
+        if self.direction is None:
+            self.direction_ = np.ones((self.n_dimensions_, ))
+        else:
+            self.direction_ = np.copy(self.direction)
+
+        validate_params({**self.get_params(), 'direction_': self.direction_,
+                         'n_dimensions_': self.n_dimensions_},
+                         self._hyperparameters)
+
+        self.direction_ = self.direction_ / np.linalg.norm(self.direction_)
+
+        axis_order = [2, 1, 3]
+        mesh_range_list = \
+            [np.arange(X.shape[order]) if self.direction_[i] >= 0
+             else -np.flip(np.arange(X.shape[order])) for i, order
+             in enumerate(axis_order[: self.n_dimensions_])]
+
+        self.mesh_ = np.stack(np.meshgrid(*mesh_range_list, indexing='xy'),
+                              axis=self.n_dimensions_)
+
+        self.max_value_ = 0.
+        self.max_value_ = np.max(self._calculate_height(
+            np.ones((1, *X.shape[1:])))) + 1
+
+        return self
+
+    def transform(self, X, y=None):
+        """For each collection of binary images, calculate the corresponding
+        collection of grayscale images based on the distance the pixel of each
+        image to the hyperplane defined by the ``direction`` vector and the
+        first seen edge of the images following that ``direction``.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Input data. Each entry along axis 0 is interpreted as a 2D or 3D
+            grayscale image.
+
+        y : None
+            There is no need of a target in a transformer, yet the pipeline API
+            requires this parameter.
+
+        Returns
+        -------
+        Xt : ndarray, shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Transformed collection of images. Each entry along axis 0 is a
+            2D or 3D binary images.
+        """
+
```

Comment:
Extra space after docstrings here, and missing one before """

Reply:
This doesn't appear to be fixed yet.

Reply:
I am not sure what you mean. flake8 does not complain. Where are those spaces supposed to be? Maybe let's discuss in person.

Reply:
Just a comment on style consistency with the rest of the library. I think flake8 only complains if there is no space after the top-level docstring for a class.

Reply:
Ahhh you mean a line! Got it :)